### ISSUE TITLE
Near command (from tag plugin) is broken

### ISSUE THREAD

**Issue by andyeff** [andyeff/skybot#2]
Error output:

```
Unhandled exception in thread started by <function run at 0x7f66d4afbb90>
Traceback (most recent call last):
  File "core/main.py", line 75, in run
    out = func(input.inp, **kw)
  File "plugins/tag.py", line 246, in near
    " and nick != lower(?) order by dist limit 20", (lat, lon, chan, nick)).fetchall()
sqlite3.OperationalError: user-defined function raised exception
```

**Comment by andyeff** [comment]
I don't know how the code works.

https://github.com/andyeff/skybot/blob/master/plugins/tag.py#L245

This makes a database call, using a database function (defined above, on line 217) that will take in two sets of latitude and longitude, and derive the distance (in kilometres) between the two points.

I don't understand how the db.execute function is apparently meant to query each user in the location table, and run the distance function against them all. I don't understand how the coords of both the user invoking the function (by typing .near) and the coords of the other people are passed into the distance function, which is apparently now not working because the error above suggests the user-defined function raised an exception - I don't know what the exception is (it doesn't tell me) but I'm going to take a guess that it didn't get the 4 parameters it was expecting.

I'm sure this has worked in the past and I've not altered this particular plugin myself so I'm puzzled why it's now broken.

**Comment by andyeff** [comment]
Huh so I think I've got some insight into this. Let's take a look at the whole line of code:

db.execute("select nick, distance(lat, lon, ?, ?) as dist from location where chan=? and nick != lower(?) order by dist limit 20", (lat, lon, chan, nick)).fetchall()

So there's two sets of lat/lon here. The first are called in the distance() function call, and THESE I believe are referring to lat/lon for the person invoking the .near command. The second set are in the second argument of the function, in a tuple with a couple other params too. This lat/lon pair is referring to the database column names in the location table, as they are being inserted as SQL parameters (where the question marks are in the first argument) to avoid SQL exploitation.

I guess my main confusion was not realising that there was a lat/lon pair of variables already in the rest of the python code. Kinda tempted to rename them so it makes it more obvious, but that might break things hilariously in ways I don't consider, so I'm gonna shove a comment in there for now to remind me.

**Comment by andyeff** [comment]
Well, big thanks to @emerencia who helped with how to trouble-shoot this. Issue is now resolved, which ultimately was fixed by just an in-place database migration / alteration, but let's go over what I learnt tonight.

* After switching the weather plugin to use a different source for weather data (Forecast.io instead of Wunderground.com), at some point coordinate data was written as strings to the database instead of floats (or REALs for sqlite3)

* This broke the near command, because instead of getting four floats passed to it, it got two or four strings instead. It wasn't written to check if the values were floats or not, so passed an exception.

So, I altered the bot's location table in the DB to have latitude and longitude be REAL instead of TEXT. Then I double checked the weather plugin was going to write floats for new users.

Now all is good.
